### ISSUE TITLE
Race condition in the janitor code of `lxc/execute` allowing to spawn persistant processes

### ISSUE THREAD

**Issue by M4GNV5** [engineer-man/piston#206]
Hey,

There is a race condition in the "janitor" which tidy ups all files and procceses after a runner exited.
The janitor code:
```bash
while pgrep -u runner$runner > /dev/null
do
    pkill -u runner$runner --signal SIGKILL
done

find /tmp -user runner$runner -delete
```

The attack goes as follows:
Spawn two runners A, and B. A creates a world executable binary and sets the suid bit. Runner B then continuosely runs that suid bit executable. This makes runner B spawn a bunch of processes as user A. If B manages to run the suid bit executable after the janitor pkill loop exits, but before the files in `/tmp` are deleted a process as user A is spawned but not killed by the janitor.

I suppose the easiest fix is to mount `/tmp` as a `tmpfs` with the `nosuid` bit? Not sure what other directories are writable by runners though and wether these can be mounted with `nosuid` too.

**Comment by Coguto** [comment]
#201 should fix it, shouldn't it?

**Comment by M4GNV5** [comment]
Not sure. The `setreuid` manual says:

```
Unprivileged users may only set the real user ID to the real user ID or the effective user ID.
```

 If i understand it correctly the suid bit executable can simply set its real user ID to user A. In the race condition case User A was already killed only user B will be killed later, so using `setreuid` to set both user ids to A means the process wont have a real id of B which is required for your patch to fix this issue.
It's just how i read the manpage though. I have an unpublished exploit for this issue, so i guess we can just test it once your PR is merged :smile:

**Comment by HexF** [comment]
Can someone give this a test on v3 and tell me how it goes?

**Comment by HexF** [comment]
@M4GNV5

**Comment by M4GNV5** [comment]
I'm having some issues getting piston to run locally:
```bash
# git pull origin master
From https://github.com/engineer-man/piston
 * branch            master     -> FETCH_HEAD
Already up to date.

# docker-compose up -d api
piston_api is up-to-date

# curl 'http://localhost:2000/runtimes'
[]

# curl 'http://localhost:2000/api/v1/packages'
<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<title>Error</title>
</head>
<body>
<pre>Cannot GET /api/v1/packages</pre>
</body>
</html>
```

The last line shows the request also done by the cli when running `cli/index.js ppman list`. Am i doing something wrong?

Anyways as v3 now seems to be merged into master(?) does this mean the officially hosted piston version is running the new v3/docker setup? I could just try the exploit there instead of having to install piston myself i guess? So I can validate the actual problem of this issue instead of fiddling with installation issues :crying_cat_face:

**Comment by dvargas46** [comment]
Yes, the public API is now using V3:
GET  https://emkc.org/api/v2/piston/runtimes
POST https://emkc.org/api/v2/piston/execute

Same with the EM discord bot.

**Comment by M4GNV5** [comment]
Okay, i did some quick tests, `/tmp` is now mounted with `nosuid`, but i could simply move to `/var/tmp` which is writeable and allows suid executables.
The programs still run in the same container without any isolation from each other.
The cleaning up of processes and files [now happens asynchronously](https://github.com/engineer-man/piston/blob/master/api/src/job.js#L235-L238), which makes it harder to reproduce the race condition. I still think its possible though, but i have to do additional testing in order to verify it.

**Comment by HexF** [comment]
You might need to run `docker-compose pull` to get a later image

**Comment by HexF** [comment]
Does this fix it?

**Comment by HexF** [comment]
I'm closing this, feel free to reopen if the issue still persists on v3.
